### DISCUSSION TITLE
关于 key , store, copy , drop 四个能力

### DISCUSSION THREAD

**Discussion by v1xingyue** [creatorsand/move-co-learn#16] [Q&A] [answered]
*  疑问1 ：
sui 里的任何一个资源都没有定义drop 能力，但是，都可以通过   object::delete(id)   删除。
https://github.com/uvd/start-sui-move/blob/main/blog_demo/sources/bolg.move#L34 

* 疑问2: 
任何一个资源都没有定义 store的能力，但是都会被存储到链上。
比如: https://github.com/uvd/start-sui-move/blob/04fb1f44e136ea0904bd6168fbb157773c1a60dd/blog_demo/sources/bolg.move#L8 

* 疑问3 :  

https://github.com/uvd/start-sui-move/tree/main 介绍的
为什么只有key 定义，就不可转移。
key + store 就可以自由转移了。

* 疑问4 : 
好像sui 里没有见过 copy 的例子，能列举个使用场景么?

**Comment by RandyPen** [comment]
object::delete(id) 是消费资源的概念，通过这种方式使用掉。drop是出了作用域自动舍弃。在编程上有区别的，譬如在闪电贷的示例代码中，限制了 借贷 是object , 无法被 drop，那就必须会还回去才能消费掉这个资源。

**Reply by fyang1024** [reply]
能不能贴一下闪电贷的实例代码的链接？谢谢

**Reply by RandyPen** [reply]
消费资源的概念，比如说，你的口袋（程序）原本是空的才能合上（结束运行），有的东西像空气，可以被drop丢弃，你合上的时候就被挤出去了。但有的东西像面包，你要进去吃完了（use) 或者拿出来(move)，才能把口袋合上。

**Comment by RandyPen** [comment, marked as answer]
具有 key 或 store 的struct 都可以存储在链上，同时具有 key, store 的可以被视为资产，使用 transfer::public_transfer 转移。而只有 key 的 struct 只能在创建该 sui object 的 module 内部被 transfer::transfer 转移。

**Comment by RandyPen** [comment]
同上，这是一种设计模式

**Comment by RandyPen** [comment]
event 类型的struct 都有 copy, drop
大部分基础数据类型都有 copy, 比如 u64, address，可以看看move语言的基础教程。啊，哈哈哈哈，虽然我不推荐看视频课学习，效率太低了，但如果确实找不到，可以去我的视频教程里翻一遍
